module/zfs/zpl_inode.c

```diff
@@ -732,10 +732,15 @@ const struct inode_operations zpl_inode_operations = {
 	.get_acl	= zpl_get_acl,
 #elif defined(HAVE_CHECK_ACL)
 	.check_acl	= zpl_check_acl,
-#elif defined(HAVE_PERMISSION)
-	.permission	= zpl_permission,
-#endif /* HAVE_GET_ACL | HAVE_CHECK_ACL | HAVE_PERMISSION */
+#endif /* HAVE_GET_ACL | HAVE_CHECK_ACL */
 #endif /* CONFIG_FS_POSIX_ACL */
+#if (defined(CONFIG_FS_POSIX_ACL) && !defined(HAVE_GET_ACL) && \
+	!defined(HAVE_CHECK_ACL)) || defined(ZFS_NFS4_ACL)
+#if defined(HAVE_PERMISSION)
```

Comment:
It looks like we can an opportunity to simplify this logic.  The `check_acl()` interface was first introduced in the 2.6.32 kernel, since this is the oldest supported kernel we can drop the `HAVE_PERMISSION` compatibility code entirely from the `CONFIG_FS_POSIX_ACL` case.  Either `get_acl` or `.check_acl` must exist.

This can then simply become the following, and `zpl_permission()` function can rely on `generic_permission(ip, mask)` for the non-nfsacl case.

```c
#if defined(ZFS_NFS4_ACL) && defined(HAVE_PERMISSION)
        .permission	= `,
#endif
```

The `HAVE_PERMISSION_WITH_NAMEIDATA` check is for an even older kernel and can be removed entirely.  Let's keep the `HAVE_PERMISSION` check though since it's already written and is useful to verify the expected prototype.

Reply:
Ok, I think I did what you wanted here, let me know if otherwise.